test/functional/autocmd/markset_spec.lua

```diff
@@ -0,0 +1,340 @@
+    command([[let g:autocmd = '']])
+    api.nvim_buf_set_lines(0, 0, -1, true, {
+      'foo\0bar',
+      'baz text',
+    })
+
+    feed('ma')
+    feed('j')
+    feed('mb')
+
+    poke_eventloop()
+
+    eq('MM', eval('g:autocmd'))
+  end)
+
+  it('is called after an uppercase mark is set', function()
+    command("autocmd MarkSet * let g:autocmd ..= 'M'")
+    command([[let g:autocmd = '']])
+    api.nvim_buf_set_lines(0, 0, -1, true, {
+      'foo\0bar',
+      'baz text',
+    })
+
+    feed('mA')
+    feed('l')
+    feed('mB')
+    feed('j')
+    feed('mC')
+
+    poke_eventloop()
+
+    eq('MMM', eval('g:autocmd'))
+  end)
+
+  it('can only be called for a specific pattern (mark)', function()
+    command("autocmd MarkSet a let g:autocmd ..= 'M'")
+    command([[let g:autocmd = '']])
+    api.nvim_buf_set_lines(0, 0, -1, true, {
+      'foo\0bar',
+      'baz text',
+    })
+
+    feed('ma')
+    feed('l')
+    feed('mb')
+    feed('j')
+    feed('ma')
+
+    poke_eventloop()
+
+    eq('MM', eval('g:autocmd'))
+  end)
+
+  it('supports glob patterns like [ab]', function()
+    command("autocmd MarkSet [ab] let g:autocmd ..= 'M'")
+    command([[let g:autocmd = '']])
+    api.nvim_buf_set_lines(0, 0, -1, true, {
+      'line 1',
+      'line 2',
+      'line 3',
+    })
+
+    feed('ma')
+    feed('j')
+    feed('mb')
+    feed('j')
+    feed('mc')
+
+    poke_eventloop()
+
+    eq('MM', eval('g:autocmd'))
+  end)
+
+  it('handles marks across multiple windows/buffers', function()
+    local orig_bufnr = api.nvim_get_current_buf()
+
+    command('enew')
+    local second_bufnr = api.nvim_get_current_buf()
+    api.nvim_buf_set_lines(second_bufnr, 0, -1, true, {
+      'second buffer line 1',
+      'second buffer line 2',
+    })
+
+    command('enew')
+    local third_bufnr = api.nvim_get_current_buf()
+    api.nvim_buf_set_lines(third_bufnr, 0, -1, true, {
+      'third buffer line 1',
+      'third buffer line 2',
+    })
+
+    command('split')
+    command('vsplit')
+
+    command('tabnew')
+    command('split')
+
+    command([[
+      let g:markset_events = []
+      autocmd MarkSet * call add(g:markset_events, {'bufnr': bufnr('%'), 'event': deepcopy(v:event)})
+    ]])
+
+    command('buffer ' .. orig_bufnr)
+    feed('gg')
+    feed('mA')
+
+    command('wincmd w')
+    command('tabnext')
+
+    feed('mB')
+
+    command('wincmd w')
+    command('enew')
+
+    local final_bufnr = api.nvim_get_current_buf()
+    api.nvim_buf_set_lines(final_bufnr, 0, -1, true, {
+      'final buffer after chaos',
+      'line 2 of final buffer',
+    })
+
+    feed('j')
+    feed('mC')
+
+    command('tabclose')
+
+    feed('mD')
+    poke_eventloop()
+
+    local mark_d_bufnr = api.nvim_get_current_buf()
+    local all_events = eval('g:markset_events')
+    eq(4, #all_events, 'Should have at 4 MarkSet events')
+
+    local mark_a = get_event(all_events, 'A')
+    eq(orig_bufnr, mark_a.bufnr)
+    eq('A', mark_a.event.name)
```

Comment:
Just test the entire thing with `eq(..., all_events)`. More readable, more meaningful, and less code.